include/openssl/hpke.h

```diff
@@ -0,0 +1,503 @@
+
+/**
+ * Suite constants, use this like:
+ *          OSSL_HPKE_SUITE myvar = OSSL_HPKE_SUITE_DEFAULT;
+ */
+# define OSSL_HPKE_SUITE_DEFAULT \
+    {\
+        OSSL_HPKE_KEM_ID_25519, \
+        OSSL_HPKE_KDF_ID_HKDF_SHA256, \
+        OSSL_HPKE_AEAD_ID_AES_GCM_128 \
+    }
+
+/**
+ * If you like your crypto turned up...
+ */
+# define OSSL_HPKE_SUITE_TURNITUPTO11 \
+    { \
+        OSSL_HPKE_KEM_ID_448, \
+        OSSL_HPKE_KDF_ID_HKDF_SHA512, \
+        OSSL_HPKE_AEAD_ID_CHACHA_POLY1305 \
+    }
+
+/**
+ * @brief HPKE single-shot encryption function
+ *
+ * This function generates an ephemeral ECDH value internally and
+ * provides the public component as an output that can be sent to
+ * the relevant private key holder along with the ciphertext.
+ *
+ * Note that the sender's public value is an output here in contrast
+ * to the case of OSSL_HPKE_enc_evp where the sender's public value
+ * is an input (along with the sender's private value).
+ *
+ * @param libctx is the context to use (normally NULL)
+ * @param mode is the HPKE mode
+ * @param suite is the ciphersuite to use
+ * @param pskid is the pskid string for a PSK mode (can be NULL)
+ * @param psklen is the psk length
+ * @param psk is the psk
+ * @param publen is the length of the public key
+ * @param pub is the encoded public key
+ * @param authprivlen is the length of the private (authentication) key
+ * @param authpriv is the encoded private (authentication) key
+ * @param authpriv_evp is the EVP_PKEY* form of private (authentication) key
+ * @param clearlen is the length of the cleartext
+ * @param clear is the encoded cleartext
+ * @param aadlen is the length of the additional data
+ * @param aad is the encoded additional data
+ * @param infolen is the length of the info data (can be zero)
+ * @param info is the encoded info data (can be NULL)
+ * @param seqlen is the length of the sequence data (can be zero)
+ * @param seq is the encoded sequence data (can be NULL)
+ * @param senderpublen length of the input buffer for sender's public key
+ * @param senderpub is the input buffer for sender public key
+ * @param cipherlen is the length of the input buffer for ciphertext
+ * @param cipher is the input buffer for ciphertext
+ * @return 1 for success, other for error (error returns can be non-zero)
+ */
+int OSSL_HPKE_enc(OSSL_LIB_CTX *libctx,
+                  unsigned int mode, OSSL_HPKE_SUITE suite,
+                  char *pskid, size_t psklen, unsigned char *psk,
+                  size_t publen, unsigned char *pub,
+                  size_t authprivlen, unsigned char *authpriv,
+                  EVP_PKEY *authpriv_evp,
+                  size_t clearlen, unsigned char *clear,
+                  size_t aadlen, unsigned char *aad,
+                  size_t infolen, unsigned char *info,
+                  size_t seqlen, unsigned char *seq,
+                  size_t *senderpublen, unsigned char *senderpub,
+                  size_t *cipherlen, unsigned char *cipher);
```

Comment:
For consistency with other existing APIs len should be always after the respective pointer argument.

Reply:
Fair enough. Did that. Could be more to do with some internal functions, will check that later.